src/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -453,37 +465,73 @@ private int InternalValueCountCore()
         /// <returns>All value names.</returns>
         private unsafe string[] GetValueNamesCore(int values)
         {
-            string[] names = new string[values];
-            char[] name = new char[MaxValueLength + 1];
-            int namelen;
+            List<string> names = new List<string>(values);
+
+            // Names in the registry aren't usually very long, although they can go to as large
+            // as 16383 characters (MaxValueLength).
+            //
+            // Every call to RegEnumValue will allocate another buffer to get the data from
+            // NtEnumerateValueKey before copying it back out to our passed in buffer. This can
+            // add up quickly- we'll try to keep the memory pressure low and grow the buffer
+            // only if needed.
 
-            fixed (char* namePtr = &name[0])
+            char[] name = ArrayPool<char>.Shared.Rent(100);
+
+            try
             {
-                for (int i = 0; i < values; i++)
+                int index = 0;
+                int result;
+                int nameLength = name.Length;
+
+                while ((result = Interop.Advapi32.RegEnumValue(
+                    _hkey,
+                    index,
```

Comment:
D'oh, I misread this as names.Length.  I can do that, but doesn't that somewhat obfuscate what is going on?